File: mcpi_sec/CryptographySession.java
package net.rozukke.elci;


import com.google.protobuf.Message;

import javax.crypto.Cipher;
import java.security.*;
import java.security.spec.*;
import java.util.Base64;

public class CryptographySession {
	public static final int keySize = 2048;
	public static final String CIPHER_INSTANCE_TYPE = "RSA";

	public final PrivateKey privateKey;
	public final PublicKey publicKey;
	public PublicKey recipientPublicKey;


	public CryptographySession() {
		System.out.println("Generating Cryptographic Key Pair");
		//Generate keys
		try {
			KeyPairGenerator keyPairGen = KeyPairGenerator.getInstance("RSA");
			keyPairGen.initialize(keySize);
			KeyPair pair = keyPairGen.generateKeyPair();

			this.privateKey = pair.getPrivate();
			this.publicKey = pair.getPublic();
		} catch (NoSuchAlgorithmException e) {
			throw new RuntimeException(e);
		}
	}

	public void setRecipientPublicKey(byte[] key){
		try {

			byte[] rawKey = Base64.getDecoder().decode(key);
			KeyFactory keyFactory = KeyFactory.getInstance("RSA");

			this.recipientPublicKey = keyFactory.generatePublic(new X509EncodedKeySpec(rawKey));
		}catch (GeneralSecurityException ex){
			ex.printStackTrace();
		}
	}

	public byte[] encrypt(byte[] data) {
		try {
			//encrypt with recipient public key
			Cipher cipherInstance = Cipher.getInstance(CIPHER_INSTANCE_TYPE);
			cipherInstance.init(Cipher.ENCRYPT_MODE, recipientPublicKey);

			byte[] out = cipherInstance.doFinal(data);
			return Base64.getEncoder().encode(out);
		}catch (GeneralSecurityException ex){
			ex.printStackTrace();
		}
		return null;
	}

	public byte[] decrypt(byte[] data){
		try {
			data = Base64.getDecoder().decode(data);
			//encrypt with recipient public key
			Cipher cipherInstance = Cipher.getInstance(CIPHER_INSTANCE_TYPE);
			cipherInstance.init(Cipher.DECRYPT_MODE, privateKey);

			return cipherInstance.doFinal(data);
		}catch (GeneralSecurityException ex){
			ex.printStackTrace();
		}
		return null;
	}

	public byte[] sign(byte[] message){

		try{

			Signature sign = Signature.getInstance("SHA256withRSA");
			sign.initSign(privateKey);
			byte[] bytes = message;
			sign.update(bytes);
			byte[] signature = sign.sign();
			return Base64.getEncoder().encode(signature);
		}catch (GeneralSecurityException ex){
			ex.printStackTrace();
		}
		return null;
	}

	public boolean verify(byte[] message, String signature){
		try{
			Signature sign = Signature.getInstance("SHA256withRSA");
			sign.initVerify(recipientPublicKey);
			byte[] bytes = message;
			sign.update(bytes);
			boolean bool = sign.verify(Base64.getDecoder().decode(signature.getBytes()));
			return bool;

		}catch (GeneralSecurityException ex){
			ex.printStackTrace();
		}
		return false;
	}




}

File: mcpi-elci/mcpi/cryptography_session.py
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_v1_5 as PKCS1_v1_5_Cipher
from Crypto import Random
from base64 import b64decode, b64encode
from hashlib import sha256
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP
from Crypto.Signature import PKCS1_v1_5
from Crypto.Hash import SHA512, SHA384, SHA256, SHA, MD5
from Crypto import Random
from base64 import b64encode, b64decode
hash = "SHA-256"


class CryptographySession:
    def __init__(self):
        print("Generating Cryptographic Key Pair")
                
        keysize = 2048
        random_generator = Random.new().read
        key = RSA.generate(keysize, random_generator)
        self.private_key, self.public_key = key, key.public_key()

    def setRecipientPublicKey(self, key):
        #key is b64 encoded
        key = b64decode(key)
        key = RSA.importKey(key)
        self.recipient_public_key = key
        pass

    def encrypt(self, data):
	    #encrypt with recipient public key
        cipher = PKCS1_v1_5_Cipher.new(self.recipient_public_key)
        data = cipher.encrypt(data)
        return b64encode(data)
    
    def decrypt(self, data):
        data = b64decode(data)
        cipher = PKCS1_v1_5_Cipher.new(self.private_key)
        return cipher.decrypt(data, None)
    
    def sign(self, message):
        signer = PKCS1_v1_5.new(self.private_key)
        digest = SHA256.new()
        digest.update(message)

        return b64encode(signer.sign(digest))
    
    def verify(self, message, signature):
        signer = PKCS1_v1_5.new(self.recipient_public_key)
        digest = SHA256.new()
        digest.update(message)
        return signer.verify(digest, b64decode(signature))